src/app/clusters/fan-control-server/fan-control-server.cpp

```diff
@@ -0,0 +1,198 @@
+#include <lib/support/CodeUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::FanControl;
+using namespace chip::app::Clusters::FanControl::Attributes;
+
+// =============================================================================
+// Pre-change callbacks for cluster attributes
+// =============================================================================
+
+using Status = Protocols::InteractionModel::Status;
+
+Protocols::InteractionModel::Status
+MatterFanControlClusterServerPreAttributeChangedCallback(const ConcreteAttributePath & attributePath,
+                                                         EmberAfAttributeType attributeType, uint16_t size, uint8_t * value)
+{
+    Protocols::InteractionModel::Status res;
+
+    switch (attributePath.mAttributeId)
+    {
+    case SpeedSetting::Id: {
+        uint8_t speedMax;
+        EmberAfStatus status = SpeedMax::Get(attributePath.mEndpointId, &speedMax);
+        VerifyOrReturnError(EMBER_ZCL_STATUS_SUCCESS == status, Status::InvalidValue);
+
+        if (*value <= speedMax)
+        {
+            res = Status::Success;
+        }
+        else
+        {
+            res = Status::InvalidValue;
+        }
+        break;
+    }
+    default:
+        res = Status::Success;
+        break;
+    }
+
+    return res;
+}
+
+void MatterFanControlClusterServerAttributeChangedCallback(const app::ConcreteAttributePath & attributePath)
+{
+    switch (attributePath.mAttributeId)
+    {
+    case FanMode::Id: {
+        FanModeType mode;
+        EmberAfStatus status = FanMode::Get(attributePath.mEndpointId, &mode);
+        VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status);
+
+        // Setting the FanMode value to Off SHALL set the values of PercentSetting, PercentCurrent,
+        // SpeedSetting, SpeedCurrent attributes to 0 (zero):.
+        if (mode == FanModeType::kOff)
+        {
+            status = PercentSetting::Set(attributePath.mEndpointId, 0);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to write PercentSetting with error: 0x%02x", status));
+
+            status = PercentCurrent::Set(attributePath.mEndpointId, 0);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to write PercentCurrent with error: 0x%02x", status));
+
+            status = SpeedSetting::Set(attributePath.mEndpointId, 0);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to write SpeedSetting with error: 0x%02x", status));
+
+            status = SpeedCurrent::Set(attributePath.mEndpointId, 0);
+            VerifyOrReturn(EMBER_ZCL_STATUS_SUCCESS == status,
+                           ChipLogError(Zcl, "Failed to write SpeedCurrent with error: 0x%02x", status));
```

Comment:
These bits are wrong.  SpeedSetting/SpeedCurrent are optional attributes.  If one of them is not enabled, the set will fail, but that should not preclude us from setting the other.

Also, it's not an error condition for these sets to fail, because these are optional attributes.

Reply:
Added feature bit condition check